tools/clang/lib/SPIRV/EmitSPIRVAction.cpp

```diff
@@ -1108,6 +1108,39 @@ class SPIRVEmitter : public ASTConsumer {
     }
   }
 
+  /// Processes the 'asfloat', 'asint', and 'asuint' intrinsic functions.
+  uint32_t processIntrinsicAsType(const CallExpr *callExpr, spv::Op toType) {
+    const QualType returnType = callExpr->getType();
+    const uint32_t returnTypeId = typeTranslator.translateType(returnType);
+    assert(callExpr->getNumArgs() == 1u);
+    const Expr *arg = callExpr->getArg(0);
+    const QualType argType = arg->getType();
+
+    // asfloat may take a float or a float vector or a float matrix as argument.
+    // These cases would be a no-op.
+    const bool returnSameScalarType = returnType == argType;
```

Comment:
I'm curious whether these three `returnSame*Type` can be replaced by a single `returnType.getCanonicalType() == arg.getCanonicalType()`? Would you mind to do a quick check? Potentially could simplify the code here. :)